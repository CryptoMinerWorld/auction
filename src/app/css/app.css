.loader {
  -webkit-animation: spin 1s linear infinite;
  animation: spin 1s linear infinite;
  border: 3px solid var(--white);
  border-top: 3px solid var(--success);
  border-radius: 50%;
  height: 75px;
  width: 75px;
}

@-webkit-keyframes spin {
  to {
    border-top-color: var(--danger);
    -webkit-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}

@keyframes spin {
  to {
    border-top-color: var(--danger);
    -webkit-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}

/* map filter */

.custom-filter-dropdown {
  padding: 8px;
  border-radius: 6px;
  background: #fff;
  -webkit-box-shadow: 0 1px 6px rgba(0, 0, 0, 0.2);
  box-shadow: 0 1px 6px rgba(0, 0, 0, 0.2);
}

.custom-filter-dropdown input {
  width: 130px;
  margin-right: 8px;
}

.custom-filter-dropdown button {
  margin-right: 8px;
}

.highlight {
  color: #f50;
}

dt {
  width: 300px;
  overflow: hidden;
  white-space: nowrap;
}
dd {
  width: 100px;
  overflow: hidden;
}

dt:after {
  content: ' ..................................................................................';
}

.card-container > .ant-tabs-card > .ant-tabs-content {
  height: auto;
  margin-top: -16px;
}

.card-container > .ant-tabs-card > .ant-tabs-content > .ant-tabs-tabpane {
  background: var(--dark-gray);
  padding: 16px;

  background-image: url(../images/gembackground.png);
  background-repeat: repeat;
  background-size: 600px;
  /* width: 100%; */
}

.card-container > .ant-tabs-card > .ant-tabs-bar {
  border-color: var(--dark-gray);
}

.card-container > .ant-tabs-card > .ant-tabs-bar .ant-tabs-tab {
  border-color: transparent;
  background: transparent;
}

.card-container > .ant-tabs-card > .ant-tabs-bar .ant-tabs-tab-active {
  border-color: var(--dark-gray);
  background: var(--dark-gray);
}
