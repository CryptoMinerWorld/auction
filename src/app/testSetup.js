import React from 'react';
import { createMemoryHistory } from "history";
import { Router } from "react-router-dom";
import {
  render
} from 'react-testing-library';

// this is a handy function that I utilize for any component
// that relies on the router being in context
export const renderWithRouter = (
    ui,
    {
      route = "/",
      history = createMemoryHistory({ initialEntries: [route] })
    } = {}
  ) => {
    return {
      ...render(<Router history={history}>{ui}</Router>),
      // adding `history` to the returned utilities to allow us
      // to reference it in our tests (just try to avoid using
      // this to test implementation details).
      history
    }}


  export const temp =  (
    ui,
    {
      route = "/",
      history = createMemoryHistory({ initialEntries: [route] })
    } = {}
  ) => ({
      ...render(<Router history={history}>{ui}</Router>),
      // adding `history` to the returned utilities to allow us
      // to reference it in our tests (just try to avoid using
      // this to test implementation details).
      history
    })

    