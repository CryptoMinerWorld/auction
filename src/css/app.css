.loader {
  -webkit-animation: spin 1s linear infinite;
  animation: spin 1s linear infinite;
  border: 3px solid var(--white);
  border-top: 3px solid var(--success);
  border-radius: 50%;
  height: 75px;
  width: 75px;
}

@-webkit-keyframes spin {
  to {
    border-top-color: var(--danger);
    -webkit-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}

@keyframes spin {
  to {
    border-top-color: var(--danger);
    -webkit-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}
