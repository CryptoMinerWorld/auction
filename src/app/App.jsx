import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { BrowserRouter } from 'react-router-dom';
import FontFaceObserver from 'fontfaceobserver';
import { connect } from 'react-redux';
import ReactGA from 'react-ga';
import Alert from 'antd/lib/alert';
import Modal from 'antd/lib/modal';
import MobileHeader from '../components/MobileHeader';
import Navbar from '../components/Nav';
import Footer from '../components/Footer';
import getWeb3 from './utils/getWeb3';
import Routes from './routes';
import './css/root.css';
import { sendContractsToRedux } from './appActions';
import { updateWalletId } from '../features/auth/authActions';
import { updatePriceOnAllLiveAuctions } from '../features/market/marketActions';
import DutchAuction from './ABI/DutchAuction.json';
import Gems from './ABI/GemERC721.json';
import Presale from './ABI/Presale.json';
require('antd/lib/alert/style/css');
require('antd/lib/modal/style/css');

// analytics breaks testing so you have to turn testmode on in development
const testMode = process.env.NODE_ENV === 'development';

ReactGA.initialize(process.env.REACT_APP_ANALYTICS, {
  testMode
});

ReactGA.pageview(window.location.pathname + window.location.search);

const dutchAuctionABI = DutchAuction.abi;
const gemsABI = Gems.abi;
const presaleABI = Presale.abi;

// @dev keeping component specific styling inside each component file is optimising for deletability. Change or delete this component in the future and all the relevant styles are removed and no more zombie css
const StickyHeader = styled.div`
  position: -webkit-sticky; /* Safari */
  position: sticky;
  top: 0;
  z-index: 3;
`;

const NotStickyHeader = styled.div`
  position: relative;
  z-index: 2;
`;

const select = store => ({
  visible: store.app.modalVisible
});

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      font: '',
      err: '',
      visible: false
    };
  }

  async componentDidMount() {
    const {
      handleSendContractsToRedux,
      handleUpdatePriceOnAllLiveAuctions,
      handleUpdateWalletId
    } = this.props;
    // @notice loading a custom font when app mounts
    const font = new FontFaceObserver('Muli', {
      weight: 400
    });
    font
      .load()
      .then(() => this.setState({ font: 'muli' }))
      .catch(error => error);

    // @notice loading web3 when component mounts
    const Web3 = await getWeb3;
    const { web3 } = Web3;

    const currentAccountId = await web3.eth
      .getAccounts()
      .then(accounts => accounts[0]);

    web3.currentProvider.publicConfigStore.on('update', ({ selectedAddress }) =>
      handleUpdateWalletId(selectedAddress)
    );

    // @notice instantiating auction contract
    const dutchContract = new web3.eth.Contract(
      dutchAuctionABI,
      process.env.REACT_APP_DUTCH_AUCTION,
      {
        from: currentAccountId
      }
    );

    const presaleContract = new web3.eth.Contract(
      presaleABI,
      process.env.REACT_APP_PRESALE,
      {
        from: currentAccountId
      }
    );

    // @notice instantiating gem contract
    const gemsContract = new web3.eth.Contract(
      gemsABI,
      process.env.REACT_APP_GEM_ERC721,
      {
        from: currentAccountId
      }
    );

    Promise.all([
      dutchContract,
      gemsContract,
      currentAccountId,
      presaleContract
    ])
      .then(
        ([
          dutchAuctionContractInstance,
          gemsContractInstance,
          currentAccount,
          presaleContract
        ]) => {
          handleSendContractsToRedux(
            dutchAuctionContractInstance,
            gemsContractInstance,
            web3,
            presaleContract,
            currentAccount
          );
        }
      )
      .catch(err => {
        this.setState({ err });
      });

    this.updatePriceOnAllLiveAuctions = setInterval(
      () => handleUpdatePriceOnAllLiveAuctions(),
      10000
    );
  }

  componentWillUnmount() {
    // @notice clear price update interval when you leav ethe app to stop any memory leaks
    clearInterval(this.priceInterval);
    clearInterval(this.updatePriceOnAllLiveAuctions);
  }

  render() {
    const { font, err } = this.state;
    const { visible } = this.props;
    return (
      <BrowserRouter>
        <main className={font}>
          {err && (
            <Alert
              message="Error Text"
              description={`${err.message}`}
              type="error"
              closable
            />
          )}
          <Modal
            visible={visible}
            title="Please Confirm Your Transaction In Metamask to Proceed"
            iconType="loading"
            zIndex={1000}
            footer={false}
            maskClosable={false}
            closable={false}
          >
            <p>
              Once you pay for the Gem using Metamask, you will be redirected to
              your workshop.
            </p>
            <strong>This may take a few moments.</strong>
          </Modal>

          <StickyHeader>
            <Navbar />
          </StickyHeader>
          <NotStickyHeader>
            <MobileHeader />
          </NotStickyHeader>
          <Routes />
          <Footer />
        </main>
      </BrowserRouter>
    );
  }
}

const actions = {
  handleSendContractsToRedux: sendContractsToRedux,
  handleUpdatePriceOnAllLiveAuctions: updatePriceOnAllLiveAuctions,
  handleUpdateWalletId: updateWalletId
};

export default connect(
  select,
  actions
)(App);

App.propTypes = {
  handleSendContractsToRedux: PropTypes.func.isRequired,
  handleUpdatePriceOnAllLiveAuctions: PropTypes.func.isRequired,
  handleUpdateWalletId: PropTypes.func.isRequired
};
