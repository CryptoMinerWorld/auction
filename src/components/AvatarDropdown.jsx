// @ts-check
import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import Menu from 'antd/lib/menu';
import Dropdown from 'antd/lib/dropdown';
import Badge from 'antd/lib/badge';
import Avatar from 'antd/lib/avatar';
import { NavLink } from 'react-router-dom';

require('antd/lib/dropdown/style/css');
require('antd/lib/badge/style/css');
require('antd/lib/menu/style/css');
require('antd/lib/avatar/style/css');

const menu = (
  <Menu>
    <Menu.Item key="0">
      <a href="http://www.alipay.com/">1st menu item</a>
    </Menu.Item>
    <Menu.Item key="1">
      <a href="http://www.taobao.com/">2nd menu item</a>
    </Menu.Item>
  </Menu>
);

/**
 * @param {{
 * to: string,
 * userImage: string
 * userName: string
 * walletId: string
 * }} AvatarDropdownProps
 */
const AvatarDropdown = ({
  to, userImage, userName, walletId,
}) => {
  const [visibility, setVisibility] = useState(false);

  useEffect(() => {
    console.log('walletId', walletId);
    return () => {
      console.log('unmounting...');
    };
  }, []);

  return (
    <NavLink
      to={to}
      className="dn dib-ns"
      onClick={() => setVisibility(false)}
      onMouseEnter={() => setVisibility(true)}
      onMouseLeave={() => setVisibility(false)}
    >
      <Dropdown overlay={menu} visible={visibility}>
        <Badge count={5}>
          <Avatar src={userImage} className="dib" />
          <p className="dib" data-testid="avatarUsername">
            {userName}
          </p>
        </Badge>
      </Dropdown>
    </NavLink>
  );
};

export default AvatarDropdown;

AvatarDropdown.propTypes = {
  to: PropTypes.string.isRequired,
  userImage: PropTypes.string.isRequired,
  userName: PropTypes.string.isRequired,
  walletId: PropTypes.string.isRequired,
};
