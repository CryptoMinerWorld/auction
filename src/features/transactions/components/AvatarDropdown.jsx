// @ts-check
import React, {useState} from 'react';
import PropTypes from 'prop-types';
import Dropdown from 'antd/lib/dropdown';
import Badge from 'antd/lib/badge';
import Avatar from 'antd/lib/avatar';
// import { NavLink } from 'react-router-dom';
import connect from "react-redux/es/connect/connect";
import Icon from "antd/lib/icon";
import img from '../../../app/images/Profile-Image-Logo-60x60.png';
import {setTransactionsSeen} from "../txActions";
import styled from "styled-components";
import {TX_CONFIRMED, TX_FAILED, TX_PENDING} from "../txConstants";

require('antd/lib/dropdown/style/css');
require('antd/lib/badge/style/css');
require('antd/lib/menu/style/css');
require('antd/lib/avatar/style/css');


const menuItemContainer = {};

const geodeTypes = {
    '0': 'Small silver geode',
    '1': 'Rotund silver geode',
    '2': 'Goldish silver geode'
};

// const generateMenuItemForTx = tx => {
//     switch (tx.txMethod) {
//         case 'SILVER_SALE':
//             switch (tx.status) {
//                 case 'COMPLETED':
//                     return (
//                       <div>
//                           <p>
//                               Silver received: {tx.receipt.events.Unboxed.returnValues.silver + " "}
//                               {tx.receipt.events.Unboxed.returnValues.gold > 0 ?
//                                 "Gold received:" + tx.receipt.events.Unboxed.returnValues.gold : ""}
//                           </p>
//                           <p>
//                               Cost: {tx.ether} ETH {tx.points > 0 ? ', ' + tx.points + 'referral' +
//                             ' points' : ''}
//                           </p>
//                       </div>
//                     );
//                 case 'PENDING':
//                     return (
//                       <div>
//                           <p>
//                               Cost: {tx.ether} ETH {tx.points > 0 ? ', ' + tx.points + 'referral' +
//                             ' points' : ''}
//                           </p>
//                       </div>
//                     )
//             }
//             break;
//         case 'GEM_UPGRADE':
//             switch (tx.status) {
//                 case 'PENDING':
//                 case 'COMPLETED':
//                     return (
//                       <div>
//                           <p>
//                               From: grade {gradeConverter(tx.gem.gradeType)}, level {tx.gem.level}
//                           </p>
//                           <p>
//                               To: grade {gradeConverter(tx.gem.gradeType + tx.gradeUp)},
//                               level {tx.gem.level + tx.levelUp}
//                           </p>
//                           <p>Cost: {tx.cost} {tx.levelUp > 0 ? 'silver' : 'gold'}</p>
//                       </div>
//                     );
//             }
//             break;
//         case 'COUPON_USE':
//             switch (tx.status) {
//                 case 'COMPLETED':
//                     return (
//                       <div>
//                           <p>
//                               Received: {geodeTypes[tx.receipt.events.CouponConsumed.returnValues.boxType] + " "}
//                           </p>
//                           <p>
//                               {tx.receipt.events.CouponConsumed.returnValues.gold > 0 ?
//                                 "Additional Gold received:" + tx.receipt.events.CouponConsumed.returnValues.gold :
//                                 ""}
//                           </p>
//                           <p>
//                               {tx.receipt.events.CouponConsumed.returnValues.silver > 0 ?
//                                 "Additional Silver received:" + tx.receipt.events.CouponConsumed.returnValues.silver :
//                                 ""}
//                           </p>
//                       </div>
//                     );
//                 case 'PENDING':
//                     return (
//                       <div>
//                           <p>
//                               Code: {tx.code}
//                           </p>
//                       </div>
//                     )
//             }
//             break;
//         case 'PLOT_SALE':
//             switch (tx.status) {
//                 case 'COMPLETED':
//                     console.log('TX::', tx.receipt);
//                     return (
//                       <div>
//                           <p>
//                           </p>
//                           <p>
//                               Cost: {tx.price} ETH
//                           </p>
//                       </div>
//                     );
//                 case 'PENDING':
//                     return (
//                       <div>
//                           <p>
//                               Cost: {tx.price} ETH
//                           </p>
//                       </div>
//                     )
//             }
//             break;
//         default:
//
//             break;
//     }
// };

const LootRow = styled.div`
    margin: 1px 0;
    text-align: left;
    font-size: 10px;
`;

const generateMenuItemForTx = tx => {
    switch (tx.event) {

        case 'Updated':
            const lootArray = tx.returnValues['loot'] || [0, 0, 0, 0, 0, 0, 0, 0, 0];
            const lootEmpty = !(lootArray.find(el => Number(el) > 0));
            return (
              <div>
                  <p>Plot {tx.returnValues['plotId']} was mined.</p>
                  <p>From {tx.returnValues['offsetFrom']} to {tx.returnValues['offsetTo']}</p>
                  {!lootEmpty && <div>Loot found:</div>}
                  {Number(lootArray[0]) > 0 &&
                  <LootRow>{lootArray[0]} Level 1 Gem{Number(lootArray[0]) > 1 ? "s" : ""}</LootRow>}
                  {Number(lootArray[1]) > 0 &&
                  <LootRow>{lootArray[1]} Level 2 Gem{Number(lootArray[1]) > 1 ? "s" : ""}</LootRow>}
                  {Number(lootArray[2]) > 0 &&
                  <LootRow>{lootArray[2]} Level 3 Gem{Number(lootArray[2]) > 1 ? "s" : ""}</LootRow>}
                  {Number(lootArray[3]) > 0 &&
                  <LootRow>{lootArray[3]} Level 4 Gem{Number(lootArray[3]) > 1 ? "s" : ""}</LootRow>}
                  {Number(lootArray[4]) > 0 &&
                  <LootRow>{lootArray[4]} Level 5 Gem{Number(lootArray[4]) > 1 ? "s" : ""}</LootRow>}
                  {Number(lootArray[5]) > 0 &&
                  <LootRow>{lootArray[5]} Piece{Number(lootArray[5]) > 1 ? "s" : ""} of Silver</LootRow>}
                  {Number(lootArray[6]) > 0 &&
                  <LootRow>{lootArray[6]} Piece{Number(lootArray[6]) > 1 ? "s" : ""} of Gold</LootRow>}
                  {Number(lootArray[7]) > 0 && <LootRow>{lootArray[7]} Artifacts</LootRow>}
                  {Number(lootArray[8]) > 0 &&
                  <LootRow>{lootArray[8]} Key{Number(lootArray[8]) > 1 ? "s" : ""}</LootRow>}
              </div>
            );
        case 'Bound':
            return (
              <div>
                  <p>Gem {tx.returnValues['gemId']} was bound to the plot {tx.returnValues['plotId']}</p>
              </div>
            );
        case 'Released':
            return (
              <div>
                  <p>Gem was released from the Plot {tx.returnValues['plotId']}</p>
              </div>
            );
        default:
            break;
    }
};

const DropdownContainer = styled.div`
    min-width: 320px;
    max-height: 500px;
    overflow: hidden auto;
    display: flex;
    flex-direction: column;
    justify-content: flex-start;
    background-color: white;
    margin-top: 20px;
`;

const TxRecordContainer = styled.div`
    position: relative;
    background-color: #ececec;
    margin: 22px 2px;
    padding: 5px 0px;
    
    &:hover {
        background-color: #e0e0e0;
        &:after {
            background-color: #e0e0e0;
        }
        &:before {
            background-color: #e0e0e0;
        }
    }
   
    &:before {
        width: 100%;
        display: block;
        height: 20px;
        clip-path: polygon(5% 0, 95% 0, 100% 100%, 0 100%);
        top: -20px;
        left: 0;
        right: 0;
        position: absolute;
        background-color: #ececec;
        content: "";
    }
    
    &:after {
        width: 100%;
        display: block;
        height: 20px;
        clip-path: polygon(0 0, 100% 0, 95% 100%, 5% 100%);
        bottom: -20px
        left: 0;
        right: 0;    
        position: absolute;
        background-color: #ececec;
        content: "show events";
        font-size: 10px;
        color: grey;
        text-align: center;
    }
`;

const TxHeader = styled.div`
`;

const TxStatus = styled.div`
    background-color: ${props => {
    switch (props.status) {
        case TX_PENDING:
            return "#fdcd14";
        case TX_CONFIRMED:
            return "green";
        case TX_FAILED:
            return "red";
    }
}};
    color: ${props => {
    switch (props.status) {
        case TX_PENDING:
            return "black";
        case TX_CONFIRMED:
            return "white";
        case TX_FAILED:
            return "white";
    }
}};
    margin: 2px 0;
    padding: 2px 5px;
    font-size: 10px;
    border-radius: 3px;
    width: 65px;
    text-align: center;
    
`;

const TxDescription = styled.div`
`;

const TxInfo = styled.div`
    padding: 5px 10px 0;
    margin-top: -12px;  
`;

const TxContractEvents = styled.div`
    display: ${props => props.expanded ? "block" : "none"};
    min-height: 30px;
    background-color: #e0e0e0;
    color: #333333;
    padding: 2px 10px;
    border-radius: 2px;
    font-size: 13px;
`;

const TxConfirmedRecord = ({tx}) => {

    const [expanded, setExpanded] = useState(false);

    return (
      <TxRecordContainer style={{cursor: "pointer"}} onClick={() => setExpanded(!expanded)}>
          <TxInfo>
              <Badge count={tx.unseen ? 1 : 0}>
                  <TxHeader>{tx.type}</TxHeader>
              </Badge>
              <TxStatus status={TX_CONFIRMED}>Confirmed</TxStatus>
              <TxDescription>{tx.description}</TxDescription>
          </TxInfo>
          <TxContractEvents expanded={expanded}>
              {tx.events && tx.events.map((event, i) => {
                  return <div style={{width: '100%'}} key={i}>{event['event']}</div>
              })}
          </TxContractEvents>
          <a
            href={`https://${process.env.REACT_APP_NETWORK}.io/tx/${tx.transactionHash}`}
            key={tx.transactionHash}
            target="_blank"
            rel="noopener noreferrer"
          >
              <Icon type="link" style={{fontSize: '24px', position: 'absolute', top: '20px', right: '0px'}}
                    className="pointer blue"/>
          </a>
      </TxRecordContainer>)
};

const menu = ({transactionHistory, pendingTransactions, failedTransactions}) => (

  <DropdownContainer>

      {!transactionHistory || !pendingTransactions || !failedTransactions ||
      (failedTransactions.length === 0 && transactionHistory.length === 0 && pendingTransactions.length === 0)
      && <div>No recent transactions</div>}
      {failedTransactions && failedTransactions.map((tx) => (
        tx.hash ?
          <TxRecordContainer key={tx.hash + 'failed'}>
              <TxInfo>
                  <Badge count={1}>
                      <TxHeader>{tx.type}</TxHeader>
                  </Badge>
                  <TxStatus status={TX_FAILED}>Failed</TxStatus>
                  <TxDescription>{tx.description}</TxDescription>
              </TxInfo>
              <a
                href={`https://${process.env.REACT_APP_NETWORK}.io/tx/${tx.hash}`}
                key={tx.hash}
                target="_blank"
                rel="noopener noreferrer"
              >
                  <Icon type="link" style={{fontSize: '24px', position: 'absolute', top: '20px', right: '0px'}}
                        className="pointer blue"/>
              </a>
          </TxRecordContainer> : ""
      ))}
      {pendingTransactions && pendingTransactions.map((tx) => (
        tx.hash ?
          <TxRecordContainer key={tx.hash + 'pending'}>
              <TxInfo>
                  <Badge count={tx.unseen ? 1 : 0}>
                      <TxHeader>{tx.type}</TxHeader>
                  </Badge>
                  <TxStatus status={TX_PENDING}>Pending</TxStatus>
                  <TxDescription>{tx.description}</TxDescription>
              </TxInfo>
              <a
                href={`https://${process.env.REACT_APP_NETWORK}.io/tx/${tx.hash}`}
                key={tx.hash}
                target="_blank"
                rel="noopener noreferrer"
              >
                  <Icon type="link" style={{fontSize: '24px', position: 'absolute', top: '20px', right: '0px'}}
                        className="pointer blue"/>
              </a>
          </TxRecordContainer> : ""
      ))}
      {transactionHistory && transactionHistory.map((tx) => (
        (tx && tx.transactionHash) ?
          <TxConfirmedRecord tx={tx} key={tx.transactionHash + 'confirmed'}/> : ""
      ))}
  </DropdownContainer>
);

/**
 * @param {{
 * to: string,
 * userImage: string
 * userName: string
 * upperCaseWalletId: string
 * }} AvatarDropdownProps
 */
class AvatarDropdown extends React.Component {

    state = {
        visibility: false,
    };

    render() {
        const {user, failedTransactions, transactionHistory, pendingTransactions} = this.props;
        const unseen = ((transactionHistory && transactionHistory.length > 0 )? Math.min(transactionHistory.findIndex(tx => tx && !tx.unseen), transactionHistory.length): 0) + +(failedTransactions ? failedTransactions.length : 0);

        return (
          user && (
            user.name && user.imageURL && user.walletId &&
            <div
              className="dib-ns"
              onMouseEnter={() => this.setState({visibility: true})}
              onMouseLeave={() => {
                  this.setState({visibility: false})
              }}
            >
                <Dropdown overlay={menu({transactionHistory, pendingTransactions, failedTransactions})}
                          visible={this.state.visibility}>
                    <>
                        <Badge count={unseen}>
                            <Avatar src={user.imageURL} className="dib h-auto v-super" size="small"/>
                        </Badge>
                        <span className="dib pl2 mb0 v-top">
                            {user.name}
                        </span>
                    </>
                </Dropdown>
            </div>
            || (user.name === 'Guest' && !user.imageURL && !user.walletId) &&
            //<Dropdown overlay={<></>} visible={this.state.visibility}>
            <div className="dib-ns">
                <Avatar src={img} className="dib"/>
                <p className="dib pl2">Guest</p>
            </div>
            //</Dropdown>
          ))
    };
}

const select = store => {
    console.log('AVATAR DROPDOWN STORE:', store);
    return {
        transactions: store.tx.transactions || [],
        unseen: store.tx.transactions ? store.tx.transactions.reduce((acc, curTx) => curTx.unseen ? acc + 1 : acc, 0) : 0,
        transactionHistory: store.tx.transactionHistory,
        pendingTransactions: store.tx.pendingTransactions,
        failedTransactions: store.tx.failedTransactions,
        //hash: store.tx.txHash,
        //txConfirmations: store.tx.txConfirmations,
        //txReceipt: store.tx.txReceipt,
        //txError: store.tx.txError,
        //txCurrentUser: store.tx.txCurrentUser,
        //txMethod: store.tx.txMethod,
        //txTokenId: store.tx.txTokenId,
    }
};

const actions = {
    handleSetTransactionsSeen: setTransactionsSeen,
};

export default connect(select)(AvatarDropdown);

AvatarDropdown.propTypes = {
    // to: PropTypes.string.isRequired,
    //userImage: PropTypes.string.isRequired,
    //userName: PropTypes.string.isRequired,
    // walletId: PropTypes.string,
    upperCaseWalletId: PropTypes.string.isRequired,
};

// AvatarDropdown.defaultProps = {
//   walletId: '',
// };
