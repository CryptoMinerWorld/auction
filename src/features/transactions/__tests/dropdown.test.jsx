import React from 'react';
import { fireEvent } from 'react-testing-library';
import AvatarDropdown from '../components/AvatarDropdown';
import { renderWithRouter } from '../../../app/testSetup';
import 'jest-dom/extend-expect';

jest.mock('react-ga');

const props = {
  to: 'string',
  userImage: 'string',
  userName: 'string',
};

test('AvatarDropdown renders correctly', () => {
  const { container } = renderWithRouter(<AvatarDropdown {...props} />, {
    route: '/',
  });
  expect(container).toMatchSnapshot();
});

test('AvatarDropdown open drop down menu on hover', () => {
  const { getByTestId, queryByTestId } = renderWithRouter(<AvatarDropdown {...props} />, {
    route: '/',
  });
  expect(queryByTestId('menu')).not.toBeInTheDocument();
  fireEvent.mouseEnter(getByTestId('avatar'));
  expect(getByTestId('menu')).toBeInTheDocument();
});

test.skip('AvatarDropdown closes dropdown when you mouseLeave', () => {});
test.skip('AvatarDropdown closes dropdown when you click', () => {});
test.skip('AvatarDropdown checks for pending tx on mount', () => {});
test.skip('AvatarDropdown unsubscribes on unmount', () => {});
test.skip('AvatarDropdown unsubscribes is axe complinat', () => {});
test.skip('AvatarDropdown unsubscribes handles errors component gracefully', () => {});
