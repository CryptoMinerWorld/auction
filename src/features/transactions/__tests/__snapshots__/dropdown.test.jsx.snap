// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvatarDropdown renders correctly 1`] = `
<div>
  <div
    class="dn dib-ns"
    data-testid="avatar"
  >
    <span
      class="ant-badge"
    >
      <span
        class="ant-avatar dib ant-avatar-circle ant-avatar-image"
      >
        <img
          src="string"
        />
      </span>
    </span>
    <p
      class="dib pl2"
      data-testid="avatarUsername"
    >
      string
    </p>
  </div>
</div>
`;
